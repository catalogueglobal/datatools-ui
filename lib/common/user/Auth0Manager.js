import auth0 from 'auth0-js'
import Auth0Lock from 'auth0-lock'
import {browserHistory} from 'react-router'
import decode from 'jwt-decode'
import moment from 'moment'
import uuidv4 from 'uuid/v4'

import { getConfigProperty, getComponentMessages, getMessage } from '../util/config'

const clientID = process.env.AUTH0_CLIENT_ID
const domain = process.env.AUTH0_DOMAIN
const auth0Client = new auth0.WebAuth({ domain, clientID })

class Auth0Manager {
  /**
   * Get an Auth0Lock instance.  If one isn't available, create it.
   * I put this here because this will get called after all the config is loaded.
   * The `this.lockOptions` variable is meant to be injected in a testing environment
   *
   * @return {Object} An auth0 lock instance
   */
  getLock () {
    if (!this.lock) {
      this.lock = new Auth0Lock(
        clientID,
        domain,
        // NOTE: The `this.lockOptions` variable is meant to be injected in a
        // testing environment.
        this.lockOptions || {
          allowSignUp: false,
          auth: {
            responseType: 'token id_token',
            params: {
              scope: 'app_metadata profile email openid user_metadata'
            },
            redirect: false
          },
          autoclose: true,
          closable: true,
          oidcConformant: true,
          languageDictionary: {
            title: getMessage(getComponentMessages('Login'), 'title')
          },
          theme: {
            logo: 'https://s3-eu-west-1.amazonaws.com/analyst-logos/conveyal-128x128.png',
            primaryColor: '#2389c9'
          }
        }
      )
    }

    return this.lock
  }

  /**
   * Use the access token (token for API calls) to get the user's profile.
   * @param  {String} accessToken auth0 access token
   * @return {Object}             auth0 profile
   */
  getProfileFromToken (accessToken) {
    return new Promise((resolve, reject) => {
      this.getLock().getUserInfo(accessToken, (err, profile) => {
        if (err) reject(err)
        else resolve(profile)
      })
    })
  }

  /**
   * Retrieves the access token from localStorage. The access token is the
   * credential used to get the user profile from Auth0.
   * @return {String} auth0 access token
   */
  getAccessToken () {
    return window.localStorage.getItem('accessToken')
  }

  /**
   * Retrieves the ID token from localStorage. The ID token is used by the
   * application in API calls as with Bearer as the Authorization header.
   * @return {String} auth0 ID token
   */
  getIdToken () {
    return window.localStorage.getItem('idToken')
  }

  hideLock () {
    this.getLock().hide()
  }

  /**
   * Push browser to login route, set up lock listeners, and show the lock
   * widget.
   *
   * @param  {Function} [onHide=noop function]
   *   A function to perform if a user closes the lock without logging in
   * @param  {Function} [push]                  A dispatch-aware action function
   * @param  {Function} receiveTokenAndProfile  A dispatch-aware action function
   * @param  {Mixed} [redirectOnSuccess=false]
   *   The string of the route to redirect to upon login success
   */
  loginWithLock ({ onHide = () => {}, push, receiveTokenAndProfile, redirectOnSuccess = false }) {
    if (!window.location.pathname.endsWith('/login')) {
      // When login is called, redirect to /login (which must be listed as an
      // Auth0 callback in client settings).
      push && push('/login')
    }
    const lock = this.getLock()

    // Set up listeners for on hide and on authenticated
    let hideFn = onHide
    lock.on('hide', () => hideFn())
    // These listeners make sure the onHide function is not called
    // after a successful login
    const clearOnHideFn = () => { hideFn = () => {} }
    const registerOnHideFn = () => { hideFn = onHide }
    lock.on('signin submit', clearOnHideFn)
    lock.on('signup submit', clearOnHideFn)
    lock.on('authorization_error', registerOnHideFn)

    // On successful authentication, handle auth result and hide lock (using
    // noop function set by 'on signin submit' listener).
    lock.on('authenticated', (authResult) => {
      this.receiveAuthResult({
        authResult,
        push,
        receiveTokenAndProfile,
        redirectOnSuccess
      })

      this.hideLock()
    })

    // Finally, show the lock widget for log in.
    lock.show()
  }

  /**
   * Log a user out from this auth0 client
   * - remove token from localStorage
   * - make a call to auth0 to logout
   */
  logout () {
    if (window.location.pathname !== '/') {
      // Push browser to root page (publicly visible) before token removal, so
      // that the application doesn't try to show the lock widget because the
      // user is no longer authenticated to view the access-restricted page.
      browserHistory.push('/')
    }
    this.removeTokens()
    const {origin} = window.location
    window.location.replace(`https://${domain}/v2/logout?returnTo=${origin}&client_id=${clientID}`)
  }

  /**
   * Recive an authResult and dispatch appropriate actions
   *
   * @param  {Object} authResult                The authResult
   * @param  {Action} [push]                    react-router-redux push
   * @param  {Action} receiveTokenAndProfile
   * @param  {String} [redirectOnSuccess=false] A possible route to redirect to upon success
   * @return {Mixed}                            actions
   */
  receiveAuthResult ({
    authResult,
    push,
    receiveTokenAndProfile,
    redirectOnSuccess
  }) {
    if (!authResult) return receiveTokenAndProfile()
    const {accessToken, idToken} = authResult
    this.setAccessToken(accessToken)
    this.setIdToken(idToken)
    // Get profile with access token and return profile and ID token to store.
    return this.getProfileFromToken(accessToken)
      .then(profile => {
        const actions = [
          receiveTokenAndProfile({
            token: idToken,
            profile
          })
        ]
        if (push && redirectOnSuccess) {
          actions.push(push(redirectOnSuccess))
        }
        return actions
      })
      .catch((err) => {
        console.error('An error occurred while trying to get the user profile', err)
        this.removeTokens()
        return receiveTokenAndProfile()
      })
  }

  /**
   * Remove both access and ID tokens stored in localStorage.
   */
  removeTokens () {
    window.localStorage.removeItem('accessToken')
    window.localStorage.removeItem('idToken')
  }

  /**
   * Attempt to silently renew the session.
   *
   * @return {Promise} if successful, returns an authResult
   */
  renewAuth () {
    return new Promise((resolve, reject) => {
      const nonce = uuidv4()
      auth0Client.renewAuth({
        audience: '',
        nonce,
        postMessageDataType: 'auth0:silent-authentication',
        redirectUri: window.location.origin + '/api/auth0-silent-callback',
        scope: 'openid email',
        usePostMessage: true
      }, (err, authResult) => {
        if (err) {
          console.log('Failed to renew log in.')
          reject(err)
        } else if (!authResult.accessToken) {
          const err = new Error('accessToken not received from auth0')
          console.log(authResult)
          reject(err)
        } else if (!nonceMatches(authResult.idToken, nonce)) {
          const err = new Error('Nonce string does not match!')
          reject(err)
        } else {
          console.log('renewed auth successfully!')
          resolve(authResult)
        }
      })
    })
  }

  /**
   * Renew session by following these steps:
   * 1. Renew the user's session if needed.
   * 2. If user token hasn't yet expired, use it.
   * 3. If token has expired, try to silently renew it.
   * 4. Logout user if session renewal fails.
   *
   * @param  {Function} receiveTokenAndProfile an action function
   * @param  {Function} logout                 an action function
   * @param  {Boolean} userIsLoggedIn
   * @return {Mixed}  Return is dependent on what actions need to happen
   */
  renewSessionIfNeeded ({
    logout,
    receiveTokenAndProfile,
    userIsLoggedIn
  }) {
    // Get the ID token if we've saved it in localStorage before
    const idToken = this.getIdToken()

    if (idToken) {
      // user has logged in before
      // check if token is expired
      if (tokenIsExpired(idToken)) {
        // FIXME: is there a need to check expiration of the access token?
        // expired, try to refresh token
        if (this.isTryingToRenewExpiredToken) return []

        this.isTryingToRenewExpiredToken = true
        // console.log('token is expired. renewing.')
        return (
          this.renewAuth()
            .then((authResult) => {
              this.isTryingToRenewExpiredToken = false
              return this.receiveAuthResult({
                authResult,
                receiveTokenAndProfile
              })
            })
            .catch((err) => {
              this.isTryingToRenewExpiredToken = false
              console.error('an error occurred while trying to log in', err)
              return logout(userIsLoggedIn)
            })
        )
      } else {
        // Token is still valid.
        // Check that enough time has passed to try again.
        const profileRefreshTime = getConfigProperty('application.profile_refresh_time')
        if (
          userIsLoggedIn &&
          (profileRefreshTime < 0 ||
            (profileRefreshTime >= 0 && this.profileExpirationTime.isAfter(moment())))
        ) {
          return []
        } else if (profileRefreshTime >= 0) {
          this.profileExpirationTime = moment().add(profileRefreshTime, 'seconds')
        }

        if (this.isTryingToGetProfileFromToken) return []

        this.isTryingToGetProfileFromToken = true

        const accessToken = this.getAccessToken()
        if (!accessToken) return logout(userIsLoggedIn)
        return this.getProfileFromToken(accessToken)
          .then((profile) => {
            this.isTryingToGetProfileFromToken = false
            return receiveTokenAndProfile({
              token: idToken,
              profile
            })
          })
          .catch((err) => {
            this.isTryingToGetProfileFromToken = false
            console.error('an error occurred while trying to get the token', err)
            return logout(userIsLoggedIn)
          })
      }
    } else {
      // user hasn't logged in before
      return logout(userIsLoggedIn)
    }
  }

  resetPassword () {
    this.getLock().show({
      allowLogin: false,
      allowSignUp: false
    })
  }

  /**
   * Store the access token in local storage. The access token is the credential
   * used to get the user profile from Auth0. More info:
   * https://auth0.com/docs/tokens/access-token
   * @param {String} token Auth0 access token
   */
  setAccessToken (token) {
    window.localStorage.setItem('accessToken', token)
  }

  /**
   * Store the ID token in local storage. The ID token is used by the
   * application in API calls as with Bearer as the Authorization header. More:
   * https://auth0.com/docs/tokens/id-token
   * @param {String} token Auth0 ID token
   */
  setIdToken (token) {
    window.localStorage.setItem('idToken', token)
  }
}

/**
 * Check that the decoded token matches the nonce.
 */
function nonceMatches (token, nonce) {
  try {
    const decoded = decode(token)
    return decoded.nonce === nonce
  } catch (err) {
    console.log(err)
    return false
  }
}

/**
 * See if a token has expired
 * @param  {String}  token The auth0 token
 * @return {Boolean}       Returns true if token is expired
 */
function tokenIsExpired (token) {
  try {
    const decoded = decode(token)
    if (!decoded.exp) {
      // shouldn't happen, but assume it doesn't expire?
      return false
    }

    return (new Date()).getTime() > decoded.exp * 1000
  } catch (err) {
    console.log(err)
    return true
  }
}

const manager = new Auth0Manager()
export default manager
