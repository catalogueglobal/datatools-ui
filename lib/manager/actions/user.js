import objectPath from 'object-path'
import createAction from 'redux-actions/lib/createAction'
import {browserHistory} from 'react-router'

import {secureFetch} from '../../common/actions'
import auth0 from '../../common/user/Auth0Manager'
import UserPermissions from '../../common/user/UserPermissions'
import {setErrorMessage} from './status'

const checkingExistingLogin = createAction('CHECKING_EXISTING_LOGIN')
// const createdPublicUser = createAction('CREATED_PUBLIC_USER')
const receiveRecentActivity = createAction('RECEIVE_USER_RECENT_ACTIVITY')
export const revokeToken = createAction('REVOKE_USER_TOKEN')
const userLoggedIn = createAction('USER_LOGGED_IN')
const userLogout = createAction('USER_LOGGED_OUT')
const setRedirectOnLogin = createAction('SET_REDIRECT_ON_LOGIN')

export function checkLogin (userIsLoggedIn) {
  return [
    checkingExistingLogin(),
    auth0.renewSessionIfNeeded({
      logout,
      receiveTokenAndProfile,
      userIsLoggedIn
    })
  ]
}

// server call
export function createPublicUser (credentials) {
  window.alert('This feature is no longer maintained')
  // const url = '/api/manager/public/user'
  // return fetchAction({
  //   next: (err, res) => {
  //     if (err) {
  //       return alert('An error occurred while trying to create the user.')
  //     }
  //     return createdPublicUser(res)
  //   },
  //   options: {
  //     body: credentials,
  //     method: 'POST'
  //   },
  //   url
  // })
}

// Recent Activity for User's Subscriptions
export function getRecentActivity (user) {
  return function (dispatch, getState) {
    const userId = user.profile.user_id
    const url = `/api/manager/secure/user/${userId}/recentactivity`
    return dispatch(secureFetch(url))
      .then(response => response.json())
      .then(activity => {
        return dispatch(receiveRecentActivity(activity))
      })
  }
}

/**
 * Login to the application with an optional URL to redirect to on a successful
 * login.
 */
export function login (redirectOnSuccess) {
  return function (dispatch, getState) {
    dispatch(setRedirectOnLogin(redirectOnSuccess))
    // Login component exists at login route.
    browserHistory.push('/login')
  }
}

/**
 * Log user out from redux store and potentially from auth0 too
 * @param  {Boolean} [logoutFromAuth0=true] Whether or not to logout of auth0
 *   This should be set to false if a user is not yet logged in to avoid
 *   an infinite loop
 * @return {Action}  The userLogout action
 */
export function logout (logoutFromAuth0 = true) {
  if (logoutFromAuth0) {
    auth0.logout()
  }
  return userLogout()
}

export function receiveTokenAndProfile (authResult) {
  // authResult can be null in case of first-time user
  if (!authResult) {
    return logout()
  }

  const {token, profile} = authResult
  return userLoggedIn({
    token,
    profile,
    permissions: new UserPermissions(profile.app_metadata.datatools)
  })
}

export function removeUserSubscription (profile, subscriptionType) {
  return function (dispatch, getState) {
    const subscriptions = profile.app_metadata.datatools.find(dt => dt.client_id === process.env.AUTH0_CLIENT_ID).subscriptions || [{type: subscriptionType, target: []}]
    const index = subscriptions.findIndex(sub => sub.type === subscriptionType)
    if (index > -1) {
      subscriptions.splice(index, 1)
    } else {
      return
    }
    return dispatch(updateUserData(profile, {subscriptions: subscriptions}))
  }
}

export function unsubscribeAll (profile) {
  return function (dispatch, getState) {
    return dispatch(updateUserData(profile, {subscriptions: []}))
  }
}

export function updateTargetForSubscription (profile, target, subscriptionType) {
  return function (dispatch, getState) {
    const subscriptions = profile.app_metadata.datatools.find(dt => dt.client_id === process.env.AUTH0_CLIENT_ID).subscriptions || [{type: subscriptionType, target: []}]
    if (subscriptions.findIndex(sub => sub.type === subscriptionType) === -1) {
      subscriptions.push({type: subscriptionType, target: []})
    }
    for (var i = 0; i < subscriptions.length; i++) {
      const sub = subscriptions[i]
      if (sub.type === subscriptionType) {
        const index = sub.target.indexOf(target)
        if (index > -1) {
          sub.target.splice(index, 1)
        } else {
          sub.target.push(target)
        }
      }
    }
    return dispatch(updateUserData(profile, {subscriptions: subscriptions}))
  }
}

// server call
export function updateUserData (user, userData) {
  return function (dispatch, getState) {
    var dtIndex = user.profile
      ? user.profile.app_metadata.datatools.findIndex(dt => dt.client_id === process.env.AUTH0_CLIENT_ID)
      : user.app_metadata.datatools.findIndex(dt => dt.client_id === process.env.AUTH0_CLIENT_ID)
    var datatools = user.profile ? user.profile.app_metadata.datatools : user.app_metadata.datatools

    if (dtIndex === -1) {
      return dispatch(setErrorMessage(`Could not find user metadata\n${JSON.stringify(user)}`))
    }

    for (var type in userData) {
      datatools[dtIndex][type] = userData[type]
    }

    var payload = {
      user_id: user.user_id,
      data: datatools
    }

    const url = '/api/manager/secure/user/' + user.user_id
    return dispatch(secureFetch(url, 'put', payload))
      .then(response => response.json())
      .then(user => {
        return user
      })
  }
}

export function updateUserMetadata (profile, props) {
  return function (dispatch, getState) {
    const CLIENT_ID = process.env.AUTH0_CLIENT_ID

    const userMetadata = profile.user_metadata || {
      lang: 'en',
      datatools: [
        {
          client_id: CLIENT_ID
        }
      ]
    }
    const clientIndex = userMetadata.datatools.findIndex(d => d.client_id === CLIENT_ID)
    for (var key in props) {
      objectPath.set(userMetadata, `datatools.${clientIndex}.${key}`, props[key])
    }
    const payload = {user_metadata: userMetadata}
    const url = `https://${process.env.AUTH0_DOMAIN}/api/v2/users/${profile.user_id}`
    return dispatch(secureFetch(url, 'PATCH', payload))
      .then(response => response.json())
      .then(user => {
        return user
      })
  }
}
