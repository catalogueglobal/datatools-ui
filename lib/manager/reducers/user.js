import update from 'react-addons-update'

import UserPermissions from '../../common/user/UserPermissions'
import UserSubscriptions from '../../common/user/UserSubscriptions'

const user = (state = {
  isCheckingLogin: true,
  token: null,
  profile: null,
  permissions: null,
  recentActivity: null,
  redirectOnSuccess: null,
  subscriptions: null
}, action) => {
  switch (action.type) {
    case 'SET_REDIRECT_ON_LOGIN':
      return update(state, {redirectOnSuccess: {$set: action.payload}})
    case 'CHECKING_EXISTING_LOGIN':
      return update(state, { isCheckingLogin: { $set: true } })
    case 'USER_LOGGED_IN':
      return update(state, {
        isCheckingLogin: { $set: false },
        token: { $set: action.payload.token },
        profile: { $set: action.payload.profile },
        permissions: { $set: new UserPermissions(action.payload.profile.app_metadata.datatools) },
        subscriptions: { $set: new UserSubscriptions(action.payload.profile.app_metadata.datatools) }
      })
    case 'REVOKE_USER_TOKEN':
      return update(state, {
        token: {$set: null}
      })
    case 'USER_LOGGED_OUT':
      return update(state, {
        isCheckingLogin: { $set: false },
        token: { $set: null },
        profile: { $set: null },
        permissions: { $set: null },
        subscriptions: { $set: null }
      })
    case 'CREATED_PUBLIC_USER':
      return update(state, {
        profile: { $set: action.payload.profile },
        permissions: { $set: new UserPermissions(action.payload.profile.app_metadata.datatools) },
        subscriptions: { $set: new UserSubscriptions(action.payload.profile.app_metadata.datatools) }
      })
    case 'RECEIVE_USER_RECENT_ACTIVITY':
      return update(state, {
        recentActivity: { $set: action.payload.activity }
      })
    default:
      return state
  }
}

export default user
