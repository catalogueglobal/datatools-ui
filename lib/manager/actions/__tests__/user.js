import jwt from 'jsonwebtoken'
import moment from 'moment'

import {checkLogin, receiveTokenAndProfile} from '../user'
import auth0 from '../../../common/user/Auth0Manager'

describe('manager > actions > user > ', () => {
  describe('checkLogin', () => {
    beforeEach(() => {
      // setup localStorage
      const storage = {}
      window.localStorage = {
        getItem: (k) => storage[k],
        removeItem: (k) => { delete storage[k] },
        setItem: (k, v) => { storage[k] = v }
      }

      // clear lock options
      auth0.lockOptions = undefined
    })

    it('should work with no token saved in localStorage', () => {
      expect(checkLogin()).toMatchSnapshot()
    })

    it('should work with expired token', (done) => {
      const expiredToken = jwt.sign({
        exp: moment().subtract(1, 'hours').unix()
      }, 'signingKey')

      // Store expired auth0 tokens in local storage
      window.localStorage.setItem('idToken', expiredToken)
      window.localStorage.setItem('accessToken', expiredToken)

      auth0.lockOptions = {
        getProfileSuccess: true,
        getProfileResult: {
          app_metadata: {
            datatools: []
          }
        }
      }

      const checkLoginResult = checkLogin()
      checkLoginResult[1]
        .then((result) => {
          delete result[0].payload.token
          expect(result).toMatchSnapshot()
          done()
        })
        .catch((err) => {
          console.error(err)
          done(err)
        })
    })

    it('should work with unexpired token', (done) => {
      const validToken = jwt.sign({
        exp: moment().add(1, 'hours').unix()
      }, 'signingKey')

      // Store auth0 tokens in local storage
      window.localStorage.setItem('idToken', validToken)
      window.localStorage.setItem('accessToken', validToken)

      auth0.lockOptions = {
        getProfileSuccess: true,
        getProfileResult: {
          app_metadata: {
            datatools: []
          }
        }
      }

      const checkLoginResult = checkLogin()
      checkLoginResult[1]
        .then((result) => {
          expect(result.payload.token).toEqual(validToken)
          delete result.payload.token
          expect(result).toMatchSnapshot()
          done()
        })
    })
  })

  describe('receiveTokenAndProfile', () => {
    it('should receive token and profile', () => {
      expect(
        receiveTokenAndProfile({
          token: 'fake-token',
          profile: {
            app_metadata: {
              datatools: []
            }
          }
        })
      ).toMatchSnapshot()
    })
  })
})
