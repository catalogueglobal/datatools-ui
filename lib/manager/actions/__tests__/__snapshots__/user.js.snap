// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager > actions > user >  checkLogin should work with expired token 1`] = `
Array [
  Object {
    "payload": Object {
      "permissions": UserPermissions {
        "appPermissionLookup": Object {},
        "orgPermissionLookup": Object {},
        "organizationLookup": Object {},
        "projectLookup": Object {},
      },
      "profile": Object {
        "app_metadata": Object {
          "datatools": Array [],
        },
      },
    },
    "type": "USER_LOGGED_IN",
  },
]
`;

exports[`manager > actions > user >  checkLogin should work with no token saved in localStorage 1`] = `
Array [
  Object {
    "type": "CHECKING_EXISTING_LOGIN",
  },
  Object {
    "type": "USER_LOGGED_OUT",
  },
]
`;

exports[`manager > actions > user >  checkLogin should work with unexpired token 1`] = `
Object {
  "payload": Object {
    "permissions": UserPermissions {
      "appPermissionLookup": Object {},
      "orgPermissionLookup": Object {},
      "organizationLookup": Object {},
      "projectLookup": Object {},
    },
    "profile": Object {
      "app_metadata": Object {
        "datatools": Array [],
      },
    },
  },
  "type": "USER_LOGGED_IN",
}
`;

exports[`manager > actions > user >  receiveTokenAndProfile should receive token and profile 1`] = `
Object {
  "payload": Object {
    "permissions": UserPermissions {
      "appPermissionLookup": Object {},
      "orgPermissionLookup": Object {},
      "organizationLookup": Object {},
      "projectLookup": Object {},
    },
    "profile": Object {
      "app_metadata": Object {
        "datatools": Array [],
      },
    },
    "token": "fake-token",
  },
  "type": "USER_LOGGED_IN",
}
`;
